# FIN 4000 — Chapter 7: Optimal Risky Portfolios
*Combining two risky assets — diversification, correlation, and the minimum-variance portfolio*

**ANSWER KEY — fully worked, instructor reference**

This notebook is the Python companion to the Chapter 7 Excel template — same data, same formulas, same numbers. Where the Excel workshop has you build a formula in a cell, this notebook has you build it in a line of Python instead.

## Step 1 — Entry Quiz *(3 minutes, individually)*

1. Why does combining two imperfectly correlated assets reduce portfolio risk below the weighted average of their individual risks?
2. What happens to the diversification benefit as the correlation between two assets approaches +1?
3. What's the difference between the minimum-variance portfolio and the optimal (tangency) risky portfolio?

---

In [1]:
import numpy as np

## Step 2 — Worked Example *(10 minutes, instructor-led)*

Combine two stocks into a portfolio. Compute the portfolio's expected return and standard deviation for a 50/50 mix.

In [2]:
E_rA, sigma_A = 0.10, 0.20
E_rB, sigma_B = 0.15, 0.30
rho = 0.20
wA, wB = 0.5, 0.5

In [3]:
E_rp = wA * E_rA + wB * E_rB
var_p = wA**2*sigma_A**2 + wB**2*sigma_B**2 + 2*wA*wB*sigma_A*sigma_B*rho
sigma_p = np.sqrt(var_p)
print(f'E(rp)={E_rp:.2%}  sigma_p={sigma_p:.2%}  (below either stock alone!)')

E(rp)=12.50%  sigma_p=19.62%  (below either stock alone!)


*E(rₚ) = 0.5(10%)+0.5(15%) = 12.5%. σₚ² = 0.25(0.04)+0.25(0.09)+2(0.5)(0.5)(0.20)(0.30)(0.2) = 0.0385 → σₚ = 19.62%. Notice this is BELOW either stock's own standalone risk (20% and 30%) — that's diversification at work.*

## Step 3 — Pair Work *(12 minutes, with a partner)*

**1.** Recompute σₚ using correlation ρ = −0.2 instead of +0.2, same weights. How much did the risk drop?

In [4]:
rho_neg = -0.2
var_p_neg = wA**2*sigma_A**2 + wB**2*sigma_B**2 + 2*wA*wB*sigma_A*sigma_B*rho_neg
sigma_p_neg = np.sqrt(var_p_neg)
print(f'sigma_p (rho=-0.2) = {sigma_p_neg:.2%}')

sigma_p (rho=-0.2) = 16.28%


**2.** Recompute σₚ using correlation ρ = 1.0 (perfectly correlated). What happened to the diversification benefit?

In [5]:
rho_one = 1.0
var_p_one = wA**2*sigma_A**2 + wB**2*sigma_B**2 + 2*wA*wB*sigma_A*sigma_B*rho_one
sigma_p_one = np.sqrt(var_p_one)
print(f'sigma_p (rho=1.0) = {sigma_p_one:.2%}  (= simple weighted average, no diversification)')

sigma_p (rho=1.0) = 25.00%  (= simple weighted average, no diversification)


**3.** Find the minimum-variance weights: wᴀ = (σʙ²−σᴀσʙρ) ÷ (σᴀ²+σʙ²−2σᴀσʙρ), using the original ρ = 0.2. What's the resulting portfolio's risk?

In [6]:
wA_min = (sigma_B**2 - sigma_A*sigma_B*rho) / (sigma_A**2 + sigma_B**2 - 2*sigma_A*sigma_B*rho)
wB_min = 1 - wA_min
sigma_min = np.sqrt(wA_min**2*sigma_A**2 + wB_min**2*sigma_B**2 + 2*wA_min*wB_min*sigma_A*sigma_B*rho)
print(f'wA_min={wA_min:.2%}  sigma_min={sigma_min:.2%}')

wA_min=73.58%  sigma_min=18.06%


## Step 4 — Python Pass *(7 minutes)*

- The portfolio return and variance formulas, built from weight/return/σ/correlation cells so any correlation can be tested
- A correlation sensitivity check: −1, −0.2, 0, +0.2, +1, watching portfolio risk change with everything else held fixed
- The minimum-variance weight formula, and the resulting portfolio's risk

In [7]:
rho_values = np.array([-1, -0.2, 0, 0.2, 1])
var_p_values = wA**2*sigma_A**2 + wB**2*sigma_B**2 + 2*wA*wB*sigma_A*sigma_B*rho_values
sigma_p_values = np.sqrt(var_p_values)
for r, s in zip(rho_values, sigma_p_values):
    print(f'rho={r:>5.1f}:  sigma_p = {s:.2%}')

rho= -1.0:  sigma_p = 5.00%
rho= -0.2:  sigma_p = 16.28%
rho=  0.0:  sigma_p = 18.03%
rho=  0.2:  sigma_p = 19.62%
rho=  1.0:  sigma_p = 25.00%


## Step 5 — Discussion *(3-5 minutes, no calculation)*

> Company A and Company B are both airlines — their stock returns are highly positively correlated. Company A and Company C (a consumer staples company) have low correlation. Which pairing gives more diversification benefit, and why does the type of business matter?